### DISCUSSION TITLE
DDRAM read_write_fn function freezes

### DISCUSSION THREAD

**Discussion by alee1234567** [polarfire-soc/polarfire-soc-documentation#126] [Bare metal embedded software]
Hello I am using the ddr demo project code and noticed that the ddr passes training and is able to run the ddr_read_write_fn function in mss_ddr.c, but when I run the function after the e51() function is launched, the code is not able to fully run the ddr_read_write_fn function and freezes. 

It freezes at this line of code in ddr_write which is called by ddr_read_write_fn
*DDR_word_ptr = DATA;
The starting address is 0x8000_000 also known as LIBERO_SETTING_DDR_32_CACHE. 
It freezes almost immediately. I cannot even step though the function afterwards. 

I often get this error message when trying to stop the program:
![image](https://user-images.githubusercontent.com/109090085/178292172-51e006c3-0151-4da3-ac80-391d50ccb4a6.png)

**Comment by mallynch** [comment]
Hi,

Just to establish a base line, can you answer the following:
You are running on the Icicle kit 
You are using the reference design ( the board comes with this design preprogrammed)
You are using the latest project from github

Does it freeze the very first time it attempts to run 
*DDR_word_ptr = DATA;
i.e. The pointer DDR_word_ptr is @0x80000000

Thanks

**Reply by alee1234567** [reply]
I am not using the Icicle kit, I am using a custom board. 
The reference design is not being used. 
I am using DDR driver "0.4.0.13." which I believe is the latest version.  Is training skipped by default in this version? Would I have to reenable it? 
Before I was using DDR driver "0.3.11"

Now I am getting a new error a few lines earlier than before, it now seems to get stuck at this declaration in ddr_write
![image](https://user-images.githubusercontent.com/109090085/178584140-4238d540-b413-4d0a-8299-b23cbf2ba25b.png)
It enters a machine trap. The mcause register is 0x6.  The operating system is FREERTOS
![image](https://user-images.githubusercontent.com/109090085/178584588-13cd943b-b167-40a0-ba49-56ac9a37fa61.png)

**Reply by mallynch** [reply]
Hi,
Can you confirm which bare metal project you are using as you mention FreeRTOS?
I though you are using "mpfs-hal-ddr-demo" as mentioned previously?
Can you confirm the MPFS HAL version. This can be found in the file platform/mpfs_hal/mpfs_hal_version.h
mcause 6 is misaligned store, I can not explain why you are getting that just now, please answer the above questions for clarity.
Thank you

**Reply by alee1234567** [reply]
Most of the files were derived from mpfs-hal-ddr-demo.  I replaced the mss_ddr.c, mss_ddr_debug.c, mss_ddr_sgii_phy_defs.c, mss_ddr_sgii_regs.c, mss_ddr_test_pattern.c and mss_nwc_init.c, and their corresponding headers files using the latest files found from  https://github.com/polarfire-soc/hart-software-services/tree/e98adeee7c6418882b303909b015691ffd93b122/baremetal/polarfire-soc-bare-metal-library 
The mpfs hal version says "#define MPFS_HAL_VERSION_MAJOR 1
#define MPFS_HAL_VERSION_MINOR 8
#define MPFS_HAL_VERSION_PATCH 134" 

Other then the files I mentioned replacing with the latest ddr driver from 4.0.13, the rest should be from the mpfs hal version mentioned here. I did not replace all the mpfs hal files since some of them are currently incompatible with the codebase I have right now.

**Comment by mallynch** [comment]
Thanks for the information. 
Could you send on the mpfs-hal-ddr-demo.lst file, the value of the mepc when the exception occurs, and the value of the stack pointer (sp) when ddr_write() is entered. If it is possible can you can send a link to your project so we can look further?

**Reply by alee1234567** [reply]
Unfortunately I cannot recreate the error with the machine trap, so I cannot obtain the mepc. 

I have received a new error where the ddr read and write functions run but all read data is returned as 0. 
![image](https://user-images.githubusercontent.com/109090085/179027912-4fd9fdb6-a55e-486a-9f0b-bde84c435b88.png)

**Comment by alee1234567** [comment]
Another update is that I tested the test_ddr function from mss_ddr.c, and the code got stuck on the pdma complete function for the non cache address at 0xC0000000
![image](https://user-images.githubusercontent.com/109090085/179067494-0fbb726b-8493-4ba7-856d-3ab6cd2bd5dd.png)

**Reply by mallynch** [reply]
Hi, 

It looks as it the underlying issue is the DDR setup.
Can you reach out to your FAE so we can fully review the ddr setting and software setup.

Thanks,
Malachy

**Reply by alee1234567** [reply]
I think I will close this thread for now, and reopen a new thread about this at another time. I decided to handle another issue beforehand. One of the issues was that I forgot to call the nwc_ddr_init function and training was not being even performed when I updated the software setting. 
It is now stuck at either the SM_VERIFY or WRITLVL stage of training.

**Reply by nitindeshpande** [reply]
@alee1234567 Did you get chance to revisit this?
